### PR TITLE
Post: Method's Arguments Part 1

### PR DESCRIPTION
**Pull request by arielj**

### CODE REVIEW

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+# => 28.26
+
+other_circle = Circle.new(4)
+puts other_circle.area
+# => 50.24
+```
+
+# Positional arguments
+
+This type of arguments get that name because the order you use matters. You can have 0 or more positional arguments, and you can have required and optional arguments.
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
This type of argument gets this name because the order you use matters. You can have 0 or more positional arguments, and you can have required and optional arguments.
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+
+# Required positional arguments
+
+This positional arguments are required when calling a method, it's not that it's required to used them when defining the method (you can have no arguments, only optionals, etc). If the method defines required parameters you have to provide an argument for each position.
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
These positional arguments are required when calling a method, it's not that it's required to used them when defining the method (you can have no arguments, only optionals, etc). If the method defines required parameters you have to provide an argument for each position.
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+```
+
+We can have as many as we want:
+
+```ruby
+def foo(arg1, arg2, arg3, arg4)
+# and more and more...
+```
+
+> Note that methods with too many arguments are a sign of a bad design since the method is probably doing too many things!
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
> Note that methods with too many arguments are a sign of bad design since the method is probably doing too many things!
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+
+# Optional positional arguments
+
+Some times we may want to allow the user to provide more input but not require them to do so. For that, we can use optional arguments for which we define a default value in case the user didn't specify one. This is particularly usefull when we want to have some default behavior for our method but give the user the option to modify that, if we just hardcode the default value inside the method we wouldn't have this flexibility.
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
Some times we may want to allow the user to provide more input but not require them to do so. For that, we can use optional arguments for which we define a default value in case the user doesn't specify one. This is particularly useful when we want to have some default behavior for our method but give the user the option to modify that. If we just hardcoded the default value inside the method we wouldn't have this flexibility.
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+foo(1)
+# => arg1 is: 1
+#    arg2 is: default value
+
+foo(1,2)
+# => arg1 is: 1
+#    arg2 is: 2
+```
+
+Usually, when we mix both types of arguments, we put the required arguments first and then the optional ones. But Ruby allows as to do weird things like:
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
Usually, when we mix both types of arguments, we put the required arguments first and then the optional ones. But Ruby allows us to do weird things like:
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+#    arg2 is: 2
+#    arg3 is: 3
+
+foo(1,2)
+# => arg1 is: 1
+#    arg2 is: default value
+#    arg3 is: 2
+```
+
+Notice that, on the second call, the second argument ends up at **arg3**. My first thought was that ruby first assigned the required arguments and then the optionals, but that's not the case, check this even more confusing example:
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
Notice that, on the second call, the second argument ends up at **arg3**. My first thought was that ruby first assigned the required arguments and then the optionals, but that's not the case. Check this even more confusing example:
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+foo(1, 2)
+arg1 is: 1
+arg2 is: "default value X"
+arg3 is: "default value Y"
+arg4 is: 2
+```
+
+What Ruby seems to do is:
+- assign left most arguments to the first parameters
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
- assign leftmost arguments to the first parameters
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+- assign right most arguments to the last parameters
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
- assign rightmost arguments to the last parameters
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+- if not enough arguments, rise an error
+- if there are unused arguments, assign them to the optional arguments in order from left to right
+
+> I would recommend to avoid this combination unless it's something with a really specific requirement since it's not too intuitive and it's hard to follow.
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
> I would recommend avoiding this combination unless it's something with a really specific requirement since it's not too intuitive and it's hard to follow.
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+
+Having optional arguments not grouped will rise an error:
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
Having optional arguments not grouped will raise an error:
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+foo
+# => arg1 is: 1
+#    arg2 is: 2
+
+foo(2)
+# => arg1 is: 2
+#    arg2 is: 4
+```
+
+The only requirement is that the arguments used in the default value was previously assigned, this won't work:
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
The only requirement is that the arguments used in the default value were previously assigned, this won't work:
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+#    arg2 is: 10 # it's multiplying 5, not 1, it uses our value, not te default one
+
+foo(5, 6)
+# => arg1 is: 5
+#    arg2 is: 6 # we are actually overriding that multiplication
+```
+
+# Variable arguments
+
+This type of optional positional arguments don't have a default value. They exist only if assigned and you can access them using a special array of arguments:
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
This type of optional positional argument doesn't have a default value. It exists only if assigned and you can access it using a special array of arguments:
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+
+
+foo(1)
+# => args is: [1]
+
+foo(1,2)
+# => args is: [1, 2]
+```
+
+You can combine this with required and optional arguments too:
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
You can combine it with required and optional arguments too:
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+#    arg2 is: 2
+#    args is: []
+
+foo(1)
+# => arg1 is: 1
+#    arg2 is: default value
+#    args is: []
+```
+
+Notice that "args" is always an array, you usually deconstruct the array or us it as is as an array or to call another method:
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
Notice that "args" is always an array. You usually deconstruct the array or use it as is as an array or to call another method:
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+  puts "arg1 in foo is: #{arg1.inspect}"
+  puts "args in foo is: #{args.inspect}"
+
+  # desconstruct the array
+  dec1, dec2, *rest = args
+  puts "dec1 is: #{dec1.inspect}"
+  puts "dec2 is: #{dec2.inspect}"
+  puts "rest is: #{rest.inspect}"
+
+  # we need the * again so ruby deconstructs the array in different
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]: @arielj I am not sure I understand this sentence.

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 16]
```diff
@@ -0,0 +1,372 @@
...
+#    rest is: [4]
+#    arg1 in bar is: 2
+#    args in bar is: [3, 4]
+```
+
+> This special parameter must always be used after the other positional arguments.
+
+# Conclusion
+
+All these options gives us some flexibility, but the order being that rigid also forces us to remember and respect the order and also remember what does each position is, which really limits us.
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
All these options gives us some flexibility, but the order being that rigid also forces us to remember and respect the order and also remember what each position is, which really limits us.
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 17]
```diff
@@ -0,0 +1,372 @@
...
+
+In the next post of this series we'll talk about keyword arguments that allows us to set arguments in any order, and on the next one we'll talk about the special block argument to inject code, some extra features like array desconstruction parameters and partially applied arguments to generate other methods.
```
Comment (ABizzinotto) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]:
```suggestion
In the next post of this series we'll talk about keyword arguments that allow us to set arguments in any order, and on the next one we'll talk about the special block argument to inject code and some extra features like array deconstruction parameters and partially applied arguments to generate other methods.
```

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown (hunk 12 above)
Comment (arielj) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: How could I write is in plural? since it's always an array of arguments (it's a single parameter, but any number of arguments)

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown (hunk 13 above)
Comment (arielj) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: Similar to the other question, it doesn't feel right to me to talk about the arguments in singular. What do you think?

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown (hunk 15 above)
Comment (arielj) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: I'll add more details and the '(splat operator)' hint so it's clear what the * means

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown (hunk 17 above)
Comment (arielj) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: hmmm I don't like having more than one 'and' with no comma, maybe something like:

"...special block argument to inject code, some extra features like array deconstruction parameters, and partially applied arguments..."

?

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+date: 2020-03-15 09:30:00
+categories: [ruby, learning]
+author: arieljuod
+---
+
+Ruby is an object oriented language where everything is an object (even methods are objects of the class Method!), so everything we need to do is done by calling methods on objects. That also means that methods have to provide a lot of flexibility because they are used everywhere.
+
+Ruby provides a lot of options to pass arguments to our methods, so we'll make this topic a series so it's not too long. We'll split the options in different categories and then break down everything with some examples and/or use cases.
+
+But first, we need to define the glosary:
```
[the remark this answers is not in the document]
Comment (arielj) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: After that I'm talking about the meaning of 'Arguments' vs 'Parameters' and 'Functions' vs 'Methods'. Maybe I just don't need that line?

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 19]
```diff
@@ -0,0 +1,372 @@
...
+layout: post
+title: Ruby Method's Arguments
+date: 2020-03-15 09:30:00
+categories: [ruby, learning]
+author: arieljuod
+---
+
+Ruby is an object oriented language where everything is an object (even methods are objects of the class Method!), so everything we need to do is done by calling methods on objects. That also means that methods have to provide a lot of flexibility because they are used everywhere.
+
+Ruby provides a lot of options to pass arguments to our methods, so we'll make this topic a series so it's not too long. We'll split the options in different categories and then break down everything with some examples and/or use cases.
```
[the remark this answers is not in the document]
Comment (arielj) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: Yes, I'm thinking of 3 posts, it would be too long fitting everything in only one

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 20]
```diff
@@ -0,0 +1,372 @@
...
+But first, we need to define the glosary:
+
+<!--more-->
+
+# Methods Vs. Functions
+
+Both methods and functions serve the same purpose: encapsulate a piece of code to reuse it with ease. And some times we use them as sinonyms, but there's a conceptual difference:
+
+- Functions return a result based only on the input, functions shouldn't depend on or modify the state of something that's not in the given inputs.
+- Methods are always executed in the context of an object, so we can call a method with no arguments but we always have access to the method's state and we can modify it.
```
[the remark this answers is not in the document]
Comment (arielj) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: Hmmm I would have to add example using another programming language since ruby doesn't have functions.

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 21]
```diff
@@ -0,0 +1,372 @@
...
+
+Another thing that's worth to differenciate is the concept of arguments and parameters. When we define a method, the parameters are part of the definition, the parameters are the variables that will contain the values that we use when calling that method. Arguments are the actual values that we use when we call a method. So, in this code:
+
+```ruby
+def my_method(foo)
+  # foo is a parameter when defining the method
+end
+
+my_method("bar")
+# "baz" is the argument we use when calling the method
```
[the remark this answers is not in the document]
Comment (arielj) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: yes, "bar"

**Comment by arielj** [comment]
> Wow! A lot of information and examples, great! Some spelling and grammar suggestions. Also, what is your plan for presenting this as a series? Since other blogs might come in between? The titles tying it together?

I'll name the 3 posts "Method's Arguments Part X" and when I write a new one I'll edit the previous to add the link of the next one so it's easier to go back and forth between them.

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown (hunk 12 above)
Comment (FionaDL) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: "These types of optional positional arguments don't have default values." Then same as you had for the second sentence.

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown (hunk 13 above)
Comment (FionaDL) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: "them"

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown (hunk 18 above)
[the remark this answers is not in the document]
Comment (FionaDL) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: Yes, I think it would be fine without that line.

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown (hunk 19 above)
[the remark this answers is not in the document]
Comment (FionaDL) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: Cool, then that's the right way to say it. Though you may want to say "We'll split the options **into** different categories"

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown (hunk 20 above)
[the remark this answers is not in the document]
Comment (FionaDL) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: Ah, I see. No worries then :)

File: _posts/2020-03-15-method-s-arguments-pt-1.markdown [hunk 22]
```diff
@@ -0,0 +1,372 @@
+---
+layout: post
+title: Ruby Method's Arguments
```
Comment (lubc) [inline _posts/2020-03-15-method-s-arguments-pt-1.markdown]: @arielj what do you think if we change the title to "Exploring Method Arguments in Ruby: Part 1"?

Comment (arielj) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: Yeah sure, it's more catchy, I'll change it tomorrow

Comment (lubc) [reply _posts/2020-03-15-method-s-arguments-pt-1.markdown]: @arielj I can change it, no worries
